pkg/middleware/middleware.go

```diff
@@ -43,60 +47,200 @@ func AuthenticationMiddleware() gin.HandlerFunc {
 			}
-
 			c.JSON(http.StatusUnauthorized, er)
 			c.Abort()
 			return
 		}
 
-		claims, ok := token.Claims.(jwt.MapClaims)
-		if !ok || !token.Valid {
+		iss, _ := unverfiedParsedToken.Issuer()
+		if iss == os.Getenv("DEFAULT_ISSUER") {
+			parsedToken, err := jwt.Parse([]byte(tokenString), jwt.WithKey(jwa.HS256(), []byte(os.Getenv("API_SECRET"))), jwt.WithValidate(true))
+			if err != nil {
+				fmt.Print(err.Error())
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "Please check your credentials and try again",
+					Error:     err.Error(),
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusUnauthorized, er)
+				c.Abort()
+				return
+			}
+
+			var user models.User
+			parsedToken.Get("user", &user)
+
+			if err := db.DB.Where(models.User{Id: user.Id}).First(&user).Error; err != nil {
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "User not found",
+					Error:     err.Error(),
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+
+				c.JSON(http.StatusUnauthorized, er)
+				c.Abort()
+				return
+			}
+			c.Set("username", *user.Username)
+			c.Set("role", *user.Userlevel)
+		} else if iss == os.Getenv("OIDC_ISSUER") {
+			if auth.Jwks == nil || os.Getenv("OIDC_USERNAME_KEY") == "" {
+				log.Print("\033[31mError: OIDC environment variables not configured properly\033[0m")
+				er := models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "Something went wrong",
+					Error:     "Something went wrong",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusInternalServerError, er)
+				c.Abort()
+				return
+			}
+
+			keyset, err := auth.Jwks.Lookup(context.Background(), os.Getenv("JWKS_URI"))
+			if err != nil {
+				log.Print("\033[31mError: Failed to create a jwk.Cache from the oidc provider's URL\033[0m")
+				er := models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "Something went wrong",
+					Error:     "Something went wrong",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusInternalServerError, er)
+				c.Abort()
+				return
+			}
+
+			parsedToken, err := jwt.Parse([]byte(tokenString), jwt.WithValidate(true), jwt.WithVerify(false))
+			if err != nil {
+				fmt.Println(err.Error())
+				er := models.LicenseError{
+					Status:    http.StatusUnauthorized,
+					Message:   "Please check your credentials and try again",
+					Error:     "Please check your credentials and try again",
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusUnauthorized, er)
+				c.Abort()
+				return
+			}
+
+			keyOptions := jws.WithKeySet(keyset, jws.WithInferAlgorithmFromKey(true))
```

Comment:
The `keyOptions` is always overridden on line 151 or 154 and thus should be removed from here.

Also, as discussed, let us not use the unsafe infer algorithm method here.

Reply:
It wont be overridden when OIDC_SIGNING_ALG is set incorrectly or when key doesnt have alg header set.

The unsafe infer algorithm will be used as the last option when the OIDC_SIGNING_ALG is not set or set incorrectly or the key doesnt have alg header.